File: BackEnd-Siukat/routes/keringanan.go
package routes

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/middlewares"
	"BackEnd-Siukat/models"
	"BackEnd-Siukat/services"
	"BackEnd-Siukat/utils"
	"net/http"
	"time"
	"fmt"

	"github.com/gin-gonic/gin"
)

func KeringananRoutes(r *gin.RouterGroup) {
	group := r.Group("/keringanan")
	srv := services.KeringananService{}

	group.GET("/all", func(c *gin.Context) {
		var ms []models.Keringanan
		config.DB.Find(&ms)
		c.JSON(http.StatusOK, ms)
	})

	authGroup := group.Group("/")
	authGroup.Use(middlewares.JwtAuth())

	authGroup.POST("/add", func(c *gin.Context) {
		var req models.Keringanan
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}
		res, err := srv.Add(req, "original")
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, res)
	})

	authGroup.PUT("/edit", func(c *gin.Context) {
		noPeserta, _ := c.Get("no_peserta")
		np := noPeserta.(string)

		req := models.Keringanan{
			Flag: c.PostForm("flag"),
		}

		var student models.CMahasiswa
		config.DB.Where("no_peserta = ?", np).First(&student)

		var existing models.Keringanan
		config.DB.Where("no_peserta = ? AND atribut = ?", np, "original").First(&existing)

		fileScan, errScan := c.FormFile("file_scan_keringanan")
		if errScan == nil {
			utils.DeleteOldFile(existing.ScanKeringanan)
			filename := fmt.Sprintf("Keringanan_%s_%s", utils.SanitizeString(student.NamaCmahasiswa), np)
			newPath, err := utils.HandleDynamicUpload(c, fileScan, student.NamaCmahasiswa, np, filename)
			if err == nil {
				req.ScanKeringanan = newPath
			}
		}
		now := time.Now()
		srv.AddLog(existing, "original", np, &now)

		res, err := srv.Edit(req, np, existing, "original")
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, res)
	})

	authGroup.GET("/get-keringanan", func(c *gin.Context) {
		noPeserta, _ := c.Get("no_peserta")
		res, err := srv.GetByLoggedIn(noPeserta.(string))
		if err != nil {
			c.JSON(http.StatusOK, models.Keringanan{})
			return
		}
		c.JSON(http.StatusOK, res)
	})

	group.GET("/get-keringanan/:no_peserta", func(c *gin.Context) {
		noPeserta := c.Param("no_peserta")
		var model models.Keringanan
		err := config.DB.Where("no_peserta = ? AND atribut = ?", noPeserta, "sanggah").First(&model).Error
		if err != nil {
			err = config.DB.Where("no_peserta = ? AND atribut = ?", noPeserta, "original").First(&model).Error
		}
		if err != nil {
			c.JSON(http.StatusOK, gin.H{"msg": "data tidak ditemukan"})
			return
		}
		c.JSON(http.StatusOK, model)
	})
}

File: ZhammDataBeLama/siukat-service-dev_snmptn_2022/routes/keringanan.js
var express = require('express');
var router = express.Router();
var Keringanan = require('../models/keringanan');
var db = require('../config/database');
var passport = require('passport');
var { keringanan } = require('../services');
var multer = require('multer')
var multerConf = require('../config/multer')
var moment = require('moment');

var timestamp = moment().format().slice(0, 19).replace('T', ' ');

var upload = multer(multerConf)

router.get('/all', function (req, res, next) {
    return db.transaction(function (t) {
        Keringanan.findAll().then(users => {
            res.send(users)
        })
    }).catch(function (err) {
        res.status(500).json(err)
    })
})

router.post('/add', passport.authenticate('jwt', { session: false }), function (req, res, next) {
    return db.transaction(function (t) {
        keringanan.add(req.body, "original").then(function (response) {
            res.json(response)
        }).catch(function (err) {
            res.status(500).json(err)
        })
    }).catch(function (err) {
        res.status(500).json(err)
    })
})

router.put('/edit',
    passport.authenticate('jwt', { session: false }),
    upload.single('file_scan_keringanan'),
    function (req, res, next) {
        return db.transaction(function (t) {
            var no_peserta = req.user.no_peserta
            Keringanan.findOne({
                where: {
                    no_peserta: no_peserta
                }
            }).then(users => {
                keringanan.addLog(users, 'original', req.user.no_peserta, timestamp).then(function (response) {
                    keringanan.edit(req, no_peserta, users, 'original').then(function (response) {
                        res.json(response)
                    }).catch(function (err) {
                        res.status(500).json(err)
                    })
                }).catch(function (err) {
                    res.status(500).json(err)
                })
            })
        }).catch(function (err) {
            res.status(500).json(err)
        })
    })

router.get('/get-keringanan', passport.authenticate('jwt', { session: false }), function (req, res) {
    return db.transaction(function (t) {
        keringanan.getByLoggedIn(req.user.no_peserta).then(function (response) {
            res.json(response)
        }).catch(function (err) {
            res.status(500).json(err)
        })
    })
})

router.get('/get-keringanan/:no_peserta', passport.authenticate('jwt', { session: false }), function (req, res) {
    return db.transaction(function (t) {
        Keringanan.findOne({
            where: {
                no_peserta: req.params.no_peserta,
                atribut: 'sanggah'
            }
        }).then(users => {
            if (users) {
                res.send(users)
            } else {
                return Keringanan.findOne({
                    where: {
                        no_peserta: req.params.no_peserta,
                        atribut: 'original'
                    }
                })
            }
        }).then(users => {
            if (users) {
                res.send(users)
            } else {
                res.send({ "msg": "data tidak ditemukan" })
            }
        })
    }).catch(function (err) {
        console.log(err)
        res.status(500).json(err)
    })
})

module.exports = router;
